codex/validation.nim

```diff
@@ -0,0 +1,98 @@
+import std/sets
+import std/tables
+import std/sequtils
+import pkg/chronos
+import pkg/chronicles
+import ./market
+import ./clock
+
+export market
+export sets
+
+type
+  Validation* = ref object
+    slots: Table[SlotId, ProofRequirements]
+    maxSlots: int
+    clock: Clock
+    market: Market
+    subscriptions: seq[Subscription]
+    running: Future[void]
+    periodicity: Periodicity
+    proofTimeout: UInt256
+  ProofRequirements = ref object
+    required: HashSet[Period]
+    submitted: HashSet[Period]
+
+proc new*(_: type Validation,
+          clock: Clock,
+          market: Market,
+          maxSlots: int): Validation =
+  Validation(clock: clock, market: market, maxSlots: maxSlots)
+
+proc slots*(validation: Validation): seq[SlotId] =
+  validation.slots.keys.toSeq
+
+proc getCurrentPeriod(validation: Validation): UInt256 =
+  return validation.periodicity.periodOf(validation.clock.now().u256)
+
+proc waitUntilNextPeriod(validation: Validation) {.async.} =
+  let period = validation.getCurrentPeriod()
+  let periodEnd = validation.periodicity.periodEnd(period)
+  await validation.clock.waitUntil(periodEnd.truncate(int64))
+
+proc subscribeSlotFilled(validation: Validation) {.async.} =
+  proc onSlotFilled(requestId: RequestId, slotIndex: UInt256) =
+    let slotId = slotId(requestId, slotIndex)
+    if not validation.slots.hasKey(slotId):
+      if validation.slots.len < validation.maxSlots:
+        validation.slots[slotId] = ProofRequirements()
+  let subscription = await validation.market.subscribeSlotFilled(onSlotFilled)
+  validation.subscriptions.add(subscription)
+
+proc removeSlotsThatHaveEnded(validation: Validation) {.async.} =
+  var ended: HashSet[SlotId]
+  for slotId in validation.slots.keys:
+    let state = await validation.market.slotState(slotId)
+    if state != SlotState.Filled:
+      ended.incl(slotId)
```

Comment:
Honestly, not a big fan of this approach 😅. It seems rather wasteful on the number of RPC calls. Instead, I suggest subscribing to events and using that to remove the slots. However, that would require adding the ﻿`SlotPaid` event into the contract as that case is not covered in the ﻿`freeSlot()` method. Or use the `SlotFreed` event in that case as well.

Reply:
Good point, it is rather wasteful. The events are tricky though. We'd not only have to subscribe to the events related to the slot, but also the events for the request. As soon as a request has failed, for instance, we'd like to stop monitoring its slots. We don't want to wait until a host has called `freeSlot()`, because it might choose not to.
I opted to accept the extra RPC calls for now, in exchange for simpler code that we can more easily adapt to changes in the design. But there is a lot of room for optimisation, that's for sure :D

Reply:
Hmm you are right about listening to the failed requests, but IMHO that is like extra 30 lines of code, don't really see it that problematic. 

Just want to raise the issue of the number of RPC calls, while it should not be something we should super optimize as there are more urgent things to solve, it would be IMHO good idea to keep it in mind and write "low-number RPC-call code" if it is possible. The reason why I raise this issue is that some providers offer free tiers where you don't have to pay for access until X amount of RPC calls. It would be good for easier adoption not to cross that line if we want to make it easy for small operators to join the network.

Reply:
Agreed, it is good to be aware of the number of RPC calls. I don't want to spend the time to rewrite the code to listen to events in this PR, because there's higher priority work. But definitely something to tackle in a follow-up PR.